### PR TITLE
Implement InvalidAuthProvider

### LINKED ISSUE #1445: Users with LDAP authentication can log in without having to enter the password
**Describe the bug**
In 10.3.3, users with LDAP authentication can log in without problems, with username and password.

In 10.3.4, users with LDAP authentication do not have the need to enter the password to log in, just the user name is enough.

The LDAP Authentication plugin does not work. showing the following error in the log.

```
[2019-06-07 19:32:01.818 +02:00] [INF] Loaded assembly "LDAP-Auth, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null" from "/var/lib/jellyfin/plugins/LDAP Authentication/LDAP-Auth.dll"
[2019-06-07 19:32:01.818 +02:00] [ERR] Error getting exported types from "LDAP-Auth, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null"
System.TypeLoadException: Method 'ChangeEasyPassword' in type 'Jellyfin.Plugin.LDAP_Auth.LdapAuthenticationProviderPlugin' from assembly 'LDAP-Auth, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at Emby.Server.Implementations.ApplicationHost.GetTypes(IEnumerable`1 assemblies)+MoveNext()
```

When trying to install it also shows an error in the log.

```
[2019-06-07 19:24:53.902 +02:00] [INF] New plugin installed: "LDAP Authentication" "2.0" Release
[2019-06-07 19:24:53.909 +02:00] [FTL] Unhandled Exception
System.TypeLoadException: Method 'ChangeEasyPassword' in type 'Jellyfin.Plugin.LDAP_Auth.LdapAuthenticationProviderPlugin' from assembly 'LDAP-Auth, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null' does not have an implementation.
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at System.Linq.Enumerable.SelectManySingleSelectorIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Emby.Server.Implementations.ApplicationHost.PluginInstalled(Object sender, GenericEventArgs`1 args)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**System (please complete the following information):**
 - OS: Debian
 - Jellyfin Version: 10.3.4
 - Reverse proxy: no

### PR DESCRIPTION
**Pull request by joshuaboniface**
**Changes**
Implements the InvalidAuthProvider, which acts as a fallback if a
configured authentication provider, e.g. LDAP, is unavailable due
to a load failure or removal. Until the user or the authentication
plugin is corrected, this will cause users with the missing provider
to be locked out, while throwing errors in the logs about the issue.

Thanks to @LogicalPhallacy for the majority of the code.

**Issues**
Fixes #1445 part 2

### CODE REVIEW

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        public void ChangeEasyPassword(User user, string newPassword, string newPasswordHash)
+        {
+            // Nothing here   
+        }
+
+        public string GetPasswordHash(User user)
+        {
+            return "";
```
Comment (joshuaboniface) [inline Emby.Server.Implementations/Library/InvalidAuthProvider.cs]: @LogicalPhallacy This is what I did differently from our work last night - this way, the user selection screen will load fine, but as expected attempting to log in as any of the invalid users fails.

File: Emby.Server.Implementations/Library/UserManager.cs [hunk 2]
```diff
@@ -400,7 +407,10 @@ private IAuthenticationProvider[] GetAuthenticationProviders(User user)
 
             if (providers.Length == 0)
             {
-                providers = new IAuthenticationProvider[] { _defaultAuthenticationProvider };
+                // this function used to assign any user without an auth provider to the default.
+                // we're going to have it use a new function now.
+                _logger.LogWarning($"The user {user.Name} was found but no Authentication Provider with ID: {user.Policy.AuthenticationProviderId} was found. Assigning user to InvalidAuthProvider temporarily");
```
Comment (cvium) [inline Emby.Server.Implementations/Library/UserManager.cs]: Let the logger handle the interpolation

Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: I redid the comment and log message using proper formatting and including a slightly nicer (IMO) wording.

File: Emby.Server.Implementations/Library/UserManager.cs [hunk 3]
```diff
@@ -400,7 +407,9 @@ private IAuthenticationProvider[] GetAuthenticationProviders(User user)
 
             if (providers.Length == 0)
             {
-                providers = new IAuthenticationProvider[] { _defaultAuthenticationProvider };
+                // Assign the user to the InvalidAuthProvider since no configured auth provider was valid/found
+                _logger.LogWarning("User {0} was found with invalid/missing Authentication Provider {1}. Assigning user to InvalidAuthProvider until this is corrected", user.Name, user.Policy.AuthenticationProviderId);
```
Comment (cvium) [inline Emby.Server.Implementations/Library/UserManager.cs]:
```suggestion
                _logger.LogWarning("User {UserName} was found with invalid/missing Authentication Provider {AuthenticationProvider}. Assigning user to InvalidAuthProvider until this is corrected", user.Name, user.Policy.AuthenticationProviderId);
```

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using MediaBrowser.Controller.Authentication;
+using MediaBrowser.Controller.Entities;
+
+namespace Emby.Server.Implementations.Library
+{
+    public class InvalidAuthProvider : IAuthenticationProvider
+    {
+        public string Name => "InvalidorMissingAuthenticationProvider";
```
Comment (cvium) [inline Emby.Server.Implementations/Library/InvalidAuthProvider.cs]:
```suggestion
        public string Name => "InvalidOrMissingAuthenticationProvider";
```

File: Emby.Server.Implementations/Library/UserManager.cs [hunk 5]
```diff
@@ -307,11 +311,14 @@ public async Task<User> AuthenticateUser(string username, string password, strin
...
                         .FirstOrDefault(i => string.Equals(username, i.Name, StringComparison.OrdinalIgnoreCase));
 
-                    var hasNewUserPolicy = authenticationProvider as IHasNewUserPolicy;
-                    if (hasNewUserPolicy != null)
+                    if (authenticationProvider.GetType() != typeof(InvalidAuthProvider))
                     {
-                        var policy = hasNewUserPolicy.GetNewUserPolicy();
-                        UpdateUserPolicy(user, policy, true);
+                        var hasNewUserPolicy = authenticationProvider as IHasNewUserPolicy;
+                        if (hasNewUserPolicy != null)
```
Comment (cvium) [inline Emby.Server.Implementations/Library/UserManager.cs]:
```suggestion
                        if (authenticationProvider is IHasNewUserPolicy hasNewUserPolicy)
```

File: Emby.Server.Implementations/Library/UserManager.cs [hunk 6]
```diff
@@ -307,11 +311,14 @@ public async Task<User> AuthenticateUser(string username, string password, strin
                     user = Users
                         .FirstOrDefault(i => string.Equals(username, i.Name, StringComparison.OrdinalIgnoreCase));
 
-                    var hasNewUserPolicy = authenticationProvider as IHasNewUserPolicy;
-                    if (hasNewUserPolicy != null)
+                    if (authenticationProvider.GetType() != typeof(InvalidAuthProvider))
                     {
-                        var policy = hasNewUserPolicy.GetNewUserPolicy();
-                        UpdateUserPolicy(user, policy, true);
+                        var hasNewUserPolicy = authenticationProvider as IHasNewUserPolicy;
```
Comment (cvium) [inline Emby.Server.Implementations/Library/UserManager.cs]:
```suggestion
```

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+{
+    public class InvalidAuthProvider : IAuthenticationProvider
+    {
+        public string Name => "InvalidOrMissingAuthenticationProvider";
+
+        public bool IsEnabled => true;
+
+        public Task<ProviderAuthenticationResult> Authenticate(string username, string password)
+        {
+            throw new Exception("User Account cannot login with this provider. The Normal provider for this user cannot be found");
```
Comment (Bond-009) [inline Emby.Server.Implementations/Library/InvalidAuthProvider.cs]: A more specific exception is preferred.

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        public Task<bool> HasPassword(User user)
+        {
+            return Task.FromResult(true);
+        }
+
+        public Task ChangePassword(User user, string newPassword)
+        {
+            return Task.FromResult(true);
```
Comment (Bond-009) [inline Emby.Server.Implementations/Library/InvalidAuthProvider.cs]:
```suggestion
            return Task.CompletedTask;
```

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        public void ChangeEasyPassword(User user, string newPassword, string newPasswordHash)
+        {
+            // Nothing here   
+        }
+
+        public string GetPasswordHash(User user)
+        {
+            return "";
```
Comment (Bond-009) [inline Emby.Server.Implementations/Library/InvalidAuthProvider.cs]:
```suggestion
            return string.Empty;
```

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        public string GetEasyPasswordHash(User user)
+        {
+            return "";
```
Comment (Bond-009) [inline Emby.Server.Implementations/Library/InvalidAuthProvider.cs]:
```suggestion
            return string.Empty;
```

File: Emby.Server.Implementations/Library/UserManager.cs [hunk 11]
```diff
@@ -307,11 +311,13 @@ public async Task<User> AuthenticateUser(string username, string password, strin
                     user = Users
                         .FirstOrDefault(i => string.Equals(username, i.Name, StringComparison.OrdinalIgnoreCase));
 
-                    var hasNewUserPolicy = authenticationProvider as IHasNewUserPolicy;
-                    if (hasNewUserPolicy != null)
+                    if (authenticationProvider.GetType() != typeof(InvalidAuthProvider))
```
Comment (Bond-009) [inline Emby.Server.Implementations/Library/UserManager.cs]: This check seems redundant.

File: Emby.Server.Implementations/Library/UserManager.cs [hunk 12]
```diff
@@ -400,7 +406,9 @@ private IAuthenticationProvider[] GetAuthenticationProviders(User user)
 
             if (providers.Length == 0)
             {
-                providers = new IAuthenticationProvider[] { _defaultAuthenticationProvider };
```
Comment (Bond-009) [inline Emby.Server.Implementations/Library/UserManager.cs]: Afaik the `_defaultPasswordResetProvider` field is now unused and can be removed.

File: Emby.Server.Implementations/Library/UserManager.cs (hunk 11 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: It's not - it prevents overwriting the on-disk `policy.xml` with the invalid provider.

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs (hunk 7 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/InvalidAuthProvider.cs]: Do we have one that fits better?

File: Emby.Server.Implementations/Library/UserManager.cs (hunk 12 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: Where is this one used?

File: Emby.Server.Implementations/Library/UserManager.cs (hunk 11 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Library/UserManager.cs]: But `InvalidAuthProvider` doesn't implement `IHasNewUserPolicy`, so it shouldn't ever overwrite, right?

File: Emby.Server.Implementations/Library/UserManager.cs (hunk 12 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: Line 429-432 I suppose?

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs (hunk 7 above)
Comment (Bond-009) [reply Emby.Server.Implementations/Library/InvalidAuthProvider.cs]: Maybe `SecurityException`? I don't really know.

File: Emby.Server.Implementations/Library/UserManager.cs (hunk 11 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: @LogicalPhallacy does it?

File: Emby.Server.Implementations/Library/InvalidAuthProvider.cs (hunk 7 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/InvalidAuthProvider.cs]: SecurityException doesn't seem to exist. I think a generic Exception is good enough, it's logging why it failed anyways.

Comment (Bond-009) [reply Emby.Server.Implementations/Library/InvalidAuthProvider.cs]: Pretty sure it exists https://github.com/jellyfin/jellyfin/blob/master/MediaBrowser.Controller/Net/SecurityException.cs
;)

Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/InvalidAuthProvider.cs]: Aah, I needed an import for that.

File: Emby.Server.Implementations/Library/UserManager.cs (hunk 11 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: Actually if it doesn't implement IHasNewUserPolicy, wouldn't that throw an exception (possibly a fatal exception) if those code tries to execute? In that case we probably do want to keep this conditional.

File: Emby.Server.Implementations/Library/UserManager.cs (hunk 12 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: Actually, I can remove like 429-432 no issue, but the "_defaultPasswordResetProvider" is used grossly on line 150 to I guess get the type? So I can't remove that.

Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: All good!

File: Emby.Server.Implementations/Library/UserManager.cs (hunk 11 above)
Comment (joshuaboniface) [reply Emby.Server.Implementations/Library/UserManager.cs]: Nope, it's good without this check. No errors or issues and it doesn't overwrite.

**Pull request merged**